README.md
![](https://github.com/zeit/hyperterm-art/blob/master/branding/HyperTerm-banner.png)

[![Build Status](https://travis-ci.org/zeit/hyperterm.svg?branch=master)](https://travis-ci.org/zeit/hyperterm)
[![Slack Channel](https://zeit-slackin.now.sh/badge.svg)](https://zeit.chat/)
[![Changelog #213](https://img.shields.io/badge/changelog-%23213-lightgrey.svg)](https://changelog.com/213)

For more details, head to: https://hyperterm.org

## Usage

You can download the latest release [here](https://hyperterm.org/#installation).

If you're on macOS, you can also use [Homebrew Cask](https://caskroom.github.io/) to download the app by running these commands:

```bash
$ brew cask update
$ brew cask install hyperterm
```

## Contribute

1. If you are running Linux, install "icnsutils", "graphicsmagick" and "xz-utils"
2. [Fork](https://help.github.com/articles/fork-a-repo/) this repository to your own GitHub account and then [clone](https://help.github.com/articles/cloning-a-repository/) it to your local device
3. Install the dependencies: `npm install`
4. Build the code and watch for changes: `npm run dev`
5. In a new tab, start the application: `npm start`

To make sure that your code works in the finished application, you can generated the binaries like that:

```bash
$ npm run pack
```

After that, you'll see the binary in the `./dist` folder!

## Related Repositories

- [Art](https://github.com/zeit/hyperterm-art)
- [Website](https://github.com/zeit/hyperterm-website)
- [Sample Extension](https://github.com/zeit/hyperpower)
- [Sample Theme](https://github.com/zeit/hyperyellow)


HISTORY.md

0.7.1 / 2016-07-26
==================

### Added
  * add a slight timeout to update the `fontSmoothing` pref (#407) [@mike-engel]

### Changed
  * session: suffix `UTF-8` to locale [@rauchg]
  * revert `--ignore-scripts` [@rauchg]
  * RPC server now extends `EventEmitter` (#406) [@jaywunder]

### Fixed
  * notifications: fix `(notes)` link [@rauchg]


0.7.0 / 2016-07-25
==================

### Added
  * implement `extend-info` for the custom `plist` [@lordgiotto]
  * add `decorateEnv` to the extensions API (#370) [@ekmartin]
  * add <kbd>ctrl</kbd>+<kbd>tab</kbd>/<kbd>ctrl</kbd>+<kbd>shift</kbd>+<kbd>tab</kbd> for tab switching (#367) [@NickChristensen]
  * add `TERM_PROGRAM` and `TERM_PROGRAM_VERSION` env vars (#350) [@tootallnate]
  * add middle click to close tab functionality (#286) [@itswil]
  * adds plugin hook for decorating the electron browser options (#310) [@danprince]
  * add support for multiple cursor shapes (#352) [@rpunkfu]
  * add `LANG` env variable (#354) [@TooTallNate]
  * implement post push and post install hooks (#328) [@freebroccolo]
  * add `.editorconfig` file (#320) [@freebroccolo]
  * add window zoom menu item (#325) [@marcbachmann]
  * add cursor actions (#217) [@marcbachmann]
  * add reducer to set cwd (#271) [@hharnisc]
  * add `homebrew cask` install method (#273, #356) [@ZakariaRidouh, @arnif]
  * add support for a `registry` configuration field (#211) [@developit]
  * add an option for a non-login shell to be run (#192) [@mike-engel, @cuth]
  * expose `getWindows` and `createWindow` to plugins (#248) [@CWSpear]

### Changed
  * use `uuid` instead of `uid2` for session ids [@albinekb]
  * reorganize dir structure and implement `electron-builder` [@rauchg]
  * move to the last term with <kbd>cmd</kbd>+<kbd>9</kbd> (#261) [@ekmartin]
  * support default width/height of browser window (#257) [@tamagokun]
  * bump `electron-prebuilt` (#359) [@freebroccolo]
  * improved linux support (#341) [@c0b41]
  * auto-updater: stop notifying and being annoying [@rauchg]
  * allow the `color` config to be an object (#193) [@mike-engel]
  * configurable cursor opacity (#76) [@hharnisc]
  * update rounded header radius (#333) [@blakeembrey]
  * specify bash to run command in for editor opening (#314) [@kyleholzinger]
  * update dependencies (#319) [@freeboccolo]
  * disable pinch zoom functionality (#326) [@marcbachmann]
  * allow opening of files in webview (#305) [@cuth]
  * modified `Info.plist` generation to enable folder drop onto dock icon (#307) [@lordgiotto]
  * dynamically change the `font-smoothing` pref (#205) [@mike-engel]
  * override the buggy `hexToRGB` implementation in `hterm` (#272) [@mauro-oto]
  * do not rely on bash-style comments for pref opening (#279) [@mfpierre]
  * close window when last tab is exited (#263) [@ekmartin]
  * bring focus to main window when clicking preferences (#235) [@leo]
  * hide scrollbar when webview is active (#209) [@marcbachmann]

### Removed
  * remove `$` from terminal commands in `README.md` (#130) [@Tyriar]
  * disable linux autoupdate (#338) [@c0b41]
  * remove unused state variable (#268) [@marcbachmann]
  * remove unnecessary comment referencing use of `!important` due to aphrodite (#223) [@conorhastings]
  * remove unnecessary event listener `teardown` on Term unmount (#242) [@developit]

### Fixed
  * fix support for `mapTermsDispatch` [@albinekb]
  * fix quit problem (#343) [@c0b41, @timneutkens]
  * fix linter issues (#321) [@freebroccolo]
  * fix prop update logic for `fontSmoothing` (#302) [@mike-engel]
  * fix installing plugins that use `node-gyp` (#291) [@dfrankland]
  * fix some documentation typos (#212) [@mauro-oto]
  * index: prevent double sesson exit (#380) [@timneutkens]
  * make sure scrollbar shim doesn't capture clicks [@rauchg]
  * make preference-opening work on all shells (#267) [@szhu]
  * make <kbd>CMD</kbd>+<kbd>K</kbd> work for all commands, even `tail` (#215) [@marcbachmann]


0.6.0 / 2016-07-17
==================

### Added
  * add a copy of the MIT license (#160) [@calinou]
  * initial Travis CI support (#107) [@amilajack]
  * provide hooks to open a new tab to the same directory (#174) [@hharnisc]

### Changed
  * notification style improvements and link to release notes upon updates [@rauchg]
  * change the version of `eslint-config-standard` from `5.3.1` to `5.3.5` (#166) [@6thmonth]
  * improved installation method (#104) [@amilajack]
  * menu improvements (#185) [@sindresorhus]

### Fixed
  * fix problems with <kbd>alt</kbd> key and special characters (e.g., å, ö, and ä) (#201) [@teemuteemu]
  * fix issue where the app icon was not showing up in Linux (#126) [@code-haven]
  * fixed typo (#152) [@radarhere]
  * fix notification message (#111) [@montogeek]
  * fix hypersolar plugin example (it does not exist) (#179) [@montogeek]
  * fix `onWindow` hook (#180) [@dfrankland]
  * make `npm install` work more reliably (#172) [@marcbachmann]
  * fix icon typo in “About” dialog (#146) [@akashnimare]


0.5.0 / 2016-07-16
==================

### Added
  * index: expose `config` and `plugins` to plugin authors in electron process [@rauchg]
  * app: expose `plugins` and `config` to window [@rauchg]

### Changed
  * plugins: improve error handling and introduce `getDecoratedConfig` [@rauchg]
  * index: notify renderer of plugins changes to reload config [@rauchg]
  * index: allow plugin authors to change electron window settings [@rauchg]
  * app: preserve class names with uglification [@rauchg]  
  * config: reload config upon `plugins change` due to decorati [@rauchg]
  * allow plugin authors to override styles by ditching `!important` [@rauchg]


0.4.5 / 2016-07-14
==================

### Changed
  * performance improvements for URL matching [@rauchg]
  * improve repaint performance for writes to the term [@rauchg]

### Fixed
  * fix issue with perf degradation related to title polling [@rauchg]
  * fix resizing when only one axis changed [@rauchg]


0.4.4 / 2016-07-14
==================


index.js
const { app, BrowserWindow, shell, Menu } = require('electron');
const createRPC = require('./rpc');
const createMenu = require('./menu');
const Session = require('./session');
const genUid = require('uid2');
const { resolve } = require('path');
const isDev = require('electron-is-dev');
const AutoUpdater = require('./auto-updater');
const toHex = require('convert-css-color-name-to-hex');

// set up config
const config = require('./config');
config.init();
const plugins = require('./plugins');

const windowSet = new Set([]);

// expose to plugins
app.config = config;
app.plugins = plugins;
app.getWindows = () => new Set([...windowSet]); // return a clone

if (isDev) {
  console.log('running in dev mode');
} else {
  console.log('running in prod mode');
}

const url = 'file://' + resolve(
  isDev ? __dirname : app.getAppPath(),
  // in prod version, we copy over index.html and dist from 'app'
  // into one dist folder to avoid unwanted files in package
  isDev ? 'app' : 'build',
  'index.html'
);

console.log('electron will open', url);

app.on('window-all-closed', () => {
  if (process.platform !== 'darwin') {
    app.quit();
  }
});

app.on('ready', () => {
  function createWindow (fn) {
    const cfg = plugins.getDecoratedConfig();
    const [width, height] = cfg.windowSize || [540, 380];

    const browserDefaults = {
      width,
      height,
      minHeight: 190,
      minWidth: 370,
      titleBarStyle: 'hidden-inset',
      title: 'HyperTerm',
      backgroundColor: toHex(cfg.backgroundColor || '#000'),
      transparent: true,
      icon: resolve(__dirname, 'static/icon.png'),
      // we only want to show when the prompt
      // is ready for user input
      show: process.env.HYPERTERM_DEBUG || isDev
    };
    const browserOptions = plugins.getDecoratedBrowserOptions(browserDefaults);

    const win = new BrowserWindow(browserOptions);

    windowSet.add(win);
    win.loadURL(url);

    const rpc = createRPC(win);
    const sessions = new Map();

    // config changes
    const cfgUnsubscribe = config.subscribe(() => {
      win.webContents.send('config change');
    });

    rpc.on('init', () => {
      win.show();
      if (fn) fn(win);

      // auto updates
      if (!isDev && process.platform !== 'linux') {
        AutoUpdater(win);
      } else {
        console.log('ignoring auto updates during dev');
      }
    });

    rpc.on('new', ({ rows = 40, cols = 100, cwd = process.env.HOME }) => {
      const shell = cfg.shell;

      initSession({ rows, cols, cwd, shell }, (uid, session) => {
        sessions.set(uid, session);
        rpc.emit('session add', {
          uid,
          shell: session.shell,
          pid: session.pty.pid
        });

        session.on('data', (data) => {
          rpc.emit('session data', { uid, data });
        });

        session.on('title', (title) => {
          rpc.emit('session title', { uid, title });
        });

        session.on('exit', () => {
          rpc.emit('session exit', { uid });
          sessions.delete(uid);
        });
      });
    });

    // TODO: this goes away when we are able to poll
    // for the title ourseleves, instead of relying
    // on Session and focus/blur to subscribe
    rpc.on('focus', ({ uid }) => {
      const session = sessions.get(uid);

      if (session) {
        session.focus();
      } else {
        console.log('session not found by', uid);
      }
    });

    rpc.on('blur', ({ uid }) => {
      const session = sessions.get(uid);

      if (session) {
        session.blur();
      } else {
        console.log('session not found by', uid);
      }
    });

    rpc.on('exit', ({ uid }) => {
      const session = sessions.get(uid);
      session && session.exit();
    });

    rpc.on('unmaximize', () => {
      win.unmaximize();
    });

    rpc.on('maximize', () => {
      win.maximize();
    });

    rpc.on('resize', ({ cols, rows }) => {
      sessions.forEach((session) => {
        session.resize({ cols, rows });
      });
    });

    rpc.on('data', ({ uid, data }) => {
      sessions.get(uid).write(data);
    });

    rpc.on('open external', ({ url }) => {
      shell.openExternal(url);
    });

    rpc.win.on('move', () => {
      rpc.emit('move');
    });

    const deleteSessions = () => {
      sessions.forEach((session, key) => {
        session.removeAllListeners();
        session.destroy();
        sessions.delete(key);
      });
    };

    // we reset the rpc channel only upon
    // subsequent refreshes (ie: F5)
    let i = 0;
    win.webContents.on('did-navigate', () => {
      if (i++) {
        deleteSessions();
      }
    });

    // expose internals to extension authors
    win.rpc = rpc;
    win.sessions = sessions;

    const load = () => {
      plugins.onWindow(win);
    };

    // load plugins
    load();

    const pluginsUnsubscribe = plugins.subscribe((err) => {
      if (!err) {
        load();
        win.webContents.send('plugins change');
      }
    });

    // the window can be closed by the browser process itself
    win.on('close', () => {
      windowSet.delete(win);
      rpc.destroy();
      deleteSessions();
      cfgUnsubscribe();
      pluginsUnsubscribe();
    });
  }

  // when opening create a new window
  createWindow();

  // expose to plugins
  app.createWindow = createWindow;

  // mac only. when the dock icon is clicked
  // and we don't have any active windows open,
  // we open one
  app.on('activate', () => {
    if (!windowSet.size) {
      createWindow();
    }
  });

  const setupMenu = () => {
    const tpl = plugins.decorateMenu(createMenu({
      createWindow,
      updatePlugins: () => {
        plugins.updatePlugins({ force: true });
      }
    }));

    Menu.setApplicationMenu(Menu.buildFromTemplate(tpl));
  };

  const load = () => {
    plugins.onApp(app);
    setupMenu();
  };

  load();
  plugins.subscribe(load);
});

function initSession (opts, fn) {
  genUid(20, (err, uid) => {
    if (err) throw err;
    fn(uid, new Session(opts));
  });
}
